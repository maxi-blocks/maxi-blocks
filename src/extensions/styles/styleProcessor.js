/**
 * Internal dependencies
 */
import {
	getCustomCssObject,
	getTransformStyles,
	getTransitionStyles,
} from './helpers';
import { getSelectorsCss } from '../../components/custom-css-control/utils';
import { getTransformSelectors } from '../../components/transform-control/utils';
import getTransformTransitionData from './transitions/getTransformTransitionData';
import transitionDefault from './transitions/transitionDefault';

/**
 * External dependencies
 */
import {
	cloneDeep,
	isEmpty,
	isEqual,
	isNil,
	isObject,
	isString,
	merge,
	mergeWith,
} from 'lodash';

const BREAKPOINTS = ['general', 'xxl', 'xl', 'l', 'm', 's', 'xs'].reverse();

const objectsCleaner = obj => {
	const response = cloneDeep(obj);

	Object.entries(response).forEach(([key, val]) => {
		if (!isObject(val) || isEmpty(val)) delete response[key];
	});

	return response;
};

const repeatedBreakpointCleaner = obj => {
	const response = cloneDeep(obj);

	BREAKPOINTS.forEach(breakpoint => {
		obj[breakpoint] &&
			Object.entries(obj[breakpoint]).forEach(([key, val]) => {
				const prevBreakpoint =
					breakpoint !== 'xl' // Ensures we jump XXL on XL breakpoint
						? BREAKPOINTS[BREAKPOINTS.indexOf(breakpoint) + 1]
						: 'general';

				if (
					obj?.[prevBreakpoint]?.[key] &&
					obj[prevBreakpoint][key] === val
				)
					delete response[breakpoint][key];
			});
	});

	return response;
};

const generalBreakpointCleaner = obj => {
	const response = cloneDeep(obj);

	Object.entries(response).forEach(([key, val]) => {
		if (key === 'general') return;

		const breakpointIndex = BREAKPOINTS.indexOf(key);
		// Is last breakpoint before general
		const isLast = cloneDeep(BREAKPOINTS)
			.splice(breakpointIndex)
			.every(breakpoint => {
				if (breakpoint === 'general' || breakpoint === key) return true;
				if (!(breakpoint in response) || isEmpty(response[breakpoint]))
					return true;

				return false;
			});

		if (isLast) {
			Object.entries(val).forEach(([prop, value]) => {
				if (
					prop !== 'label' &&
					'general' in response &&
					!isNil(response.general[prop]) &&
					isEqual(value, response.general[prop])
				)
					delete response[key][prop];
			});
		}
	});

	return response;
};

const hoverStylesCleaner = (normalObj, hoverObj) => {
	if (normalObj)
		Object.entries(normalObj).forEach(([key, val]) => {
			if (hoverObj && key in hoverObj)
				Object.entries(val).forEach(([breakpoint, breakpointVal]) => {
					if (breakpoint in hoverObj[key])
						Object.entries(breakpointVal).forEach(
							([attrKey, attrVal]) => {
								// First higher breakpoint that has this property defined
								const prevBreakpoint = BREAKPOINTS.slice(
									BREAKPOINTS.indexOf(breakpoint) + 1
								).filter(
									prevBreakpoint =>
										!isNil(
											hoverObj[key]?.[prevBreakpoint]?.[
												attrKey
											]
										)
								)?.[0];

								// If higher breakpoint has hover value that is
								// different from this breakpoint hover value then keep it
								if (
									prevBreakpoint &&
									hoverObj[key]?.[prevBreakpoint]?.[
										attrKey
									] !== hoverObj[key][breakpoint][attrKey]
								) {
									return;
								}

								if (
									attrKey in hoverObj[key][breakpoint] &&
									hoverObj[key][breakpoint][attrKey] ===
										attrVal &&
									attrKey !== 'transition'
								)
									delete hoverObj[key][breakpoint][attrKey];
							}
						);
				});
		});

	return hoverObj;
};

const styleCleaner = styles => {
	Object.entries(styles).forEach(item => {
		const [target, rawVal] = item;

		// Clean non-object and empty targets
		// First clean, avoids unnecessary work on next loops
		const val = objectsCleaner(rawVal);

		if (isEmpty(val)) {
			delete styles[target];

			return;
		}

		styles[target] = val;

		// Clean breakpoint repeated values
		Object.entries(val).forEach(([typeKey, typeVal]) => {
			if (Object.keys(typeVal).length > 1)
				styles[target][typeKey] = repeatedBreakpointCleaner(typeVal);
		});

		// Clean non-necessary breakpoint values when is same than general
		Object.entries(val).forEach(([typeKey, typeVal]) => {
			if (Object.keys(typeVal).length > 1)
				styles[target][typeKey] = generalBreakpointCleaner(typeVal);
		});

		// Clean hover values
		if (target.includes(':hover')) {
			const normalKey = target.replace(':hover', '');

			styles[target] = hoverStylesCleaner(styles[normalKey], val);
		}

		// Clean empty breakpoints
		Object.entries(val).forEach(([typeKey, typeVal]) => {
			Object.entries(typeVal).forEach(([breakpoint, breakpointVal]) => {
				if (!isObject(breakpointVal) || isEmpty(breakpointVal))
					delete styles[target][typeKey][breakpoint];
			});
		});

		// Clean non-object and empty targets
		// Second clean before returning
		const cleanedVal = objectsCleaner(styles[target]);

		if (isEmpty(cleanedVal)) {
			delete styles[target];

			return;
		}

		styles[target] = cleanedVal;
	});

	return styles;
};

const styleProcessor = (obj, data, props) => {
	const selectors = data?.customCss?.selectors;
	const transitionSelectors = {
		...(data?.transition || transitionDefault),
		transform: getTransformTransitionData(selectors, props),
	};

	const styles = cloneDeep(obj);

	const transitionObject = getTransitionStyles(props, transitionSelectors);
	if (!isEmpty(transitionObject)) merge(styles, transitionObject);

	// Process custom styles if they exist
	const newCssSelectors = getSelectorsCss(selectors, props);
	const newTransformSelectors = getTransformSelectors(selectors, props);

	if (!isEmpty(newCssSelectors)) {
		const customCssObject = getCustomCssObject(newCssSelectors, props);
		!isEmpty(customCssObject) && merge(styles, customCssObject);
	}
	if (!isEmpty(newTransformSelectors)) {
		const transformObject = getTransformStyles(
			props,
			newTransformSelectors
		);

		if (!isEmpty(transformObject)) {
			const isTransformString = string =>
				isString(string) &&
				['rotate', 'scale', 'translate'].some(word =>
					string.includes(word)
				);

			// eslint-disable-next-line consistent-return
			mergeWith(styles, transformObject, (objValue, srcValue) => {
				if ([objValue, srcValue].every(isTransformString))
					return `${objValue} ${srcValue}`;
			});
		}
	}
	return styleCleaner(styles);
};

export default styleProcessor;
