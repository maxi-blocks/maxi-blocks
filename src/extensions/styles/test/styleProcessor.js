import styleProcessor from '../styleProcessor';

jest.mock('@wordpress/blocks', () => jest.fn());
jest.mock('src/components/block-inserter/index.js', () => jest.fn());
jest.mock('../../attributes/getBlockData.js', () => jest.fn());
jest.mock('src/components/transform-control/utils.js', () => ({
	getTransformSelectors: jest.fn(),
}));

describe('styleCleaner', () => {
	it('Returns cleaned styles obj', () => {
		const obj = {
			'': {
				margin: {
					general: {},
					xxl: {},
					xl: {},
					l: {},
					m: {},
					s: {},
					xs: {},
				},
				padding: {
					general: {},
					xxl: {},
					xl: {},
					l: {},
					m: {},
					s: {},
					xs: {},
				},
				border: {
					general: {
						'border-color': 'var(--maxi-light-color-2)',
						'border-style': 'solid',
						'border-top-width': '2px',
						'border-right-width': '2px',
						'border-bottom-width': '2px',
						'border-left-width': '2px',
					},
					xxl: {},
					xl: {},
					l: {},
					m: {},
					s: {},
					xs: {},
				},
				size: {
					general: {},
					xxl: {},
					xl: {},
					l: {},
					m: {},
					s: {},
					xs: {},
				},
				boxShadow: {
					general: {
						'box-shadow':
							'0px 0px 0px 0px var(--maxi-light-color-1)',
					},
				},
				opacity: {
					general: {},
					xxl: {},
					xl: {},
					l: {},
					m: {},
					s: {},
					xs: {},
				},
				zIndex: {},
				position: {},
				display: {},
				transform: {},
			},
			':hover': {
				border: {
					general: {
						'border-color': 'var(--maxi-light-color-6)',
						'border-style': 'solid',
						'border-top-width': '2px',
						'border-right-width': '2px',
						'border-bottom-width': '2px',
						'border-left-width': '2px',
					},
					xxl: {},
					xl: {},
					l: {},
					m: {},
					s: {},
					xs: {},
				},
				boxShadow: {
					general: {
						'box-shadow':
							'5px 5px 0px 0px var(--maxi-light-color-6)',
					},
				},
			},
			' > .maxi-background-displayer': {
				border: {
					general: {
						'border-color': 'var(--maxi-light-color-2)',
						'border-style': 'solid',
						'border-top-width': '2px',
						'border-right-width': '2px',
						'border-bottom-width': '2px',
						'border-left-width': '2px',
					},
					xxl: {},
					xl: {},
					l: {},
					m: {},
					s: {},
					xs: {},
				},
			},
			':hover > .maxi-background-displayer': {
				border: {
					general: {
						'border-color': 'var(--maxi-light-color-6)',
						'border-style': 'solid',
						'border-top-width': '2px',
						'border-right-width': '2px',
						'border-bottom-width': '2px',
						'border-left-width': '2px',
					},
					xxl: {},
					xl: {},
					l: {},
					m: {},
					s: {},
					xs: {},
				},
				size: {
					general: {
						transform:
							'translate(calc(-50% - 0px), calc(-50% - 0px))',
					},
					xxl: {
						transform:
							'translate(calc(-50% - 0px), calc(-50% - 0px))',
					},
					xl: {
						transform:
							'translate(calc(-50% - 0px), calc(-50% - 0px))',
					},
					l: {
						transform:
							'translate(calc(-50% - 0px), calc(-50% - 0px))',
					},
					m: {
						transform:
							'translate(calc(-50% - 0px), calc(-50% - 0px))',
					},
					s: {
						transform:
							'translate(calc(-50% - 0px), calc(-50% - 0px))',
					},
					xs: {
						transform:
							'translate(calc(-50% - 0px), calc(-50% - 0px))',
					},
				},
			},
			' .maxi-container-arrow': {
				shadow: {
					general: {
						filter: 'drop-shadow(0px 0px 0px var(--maxi-light-color-1))',
					},
				},
			},
			' .maxi-container-arrow .maxi-container-arrow--content': {
				arrow: {},
			},
			' .maxi-container-arrow .maxi-container-arrow--content:after': {
				background: {
					label: 'Arrow Color',
					general: {},
				},
			},
			' .maxi-container-arrow .maxi-container-arrow--content:before': {
				border: {
					label: 'Arrow Border',
					general: {
						background: 'var(--maxi-undefined-color-2)',
						top: 'calc(1px)',
						left: 'calc(1px)',
						width: 'calc(50% + 4px)',
						height: 'calc(50% + 4px)',
					},
				},
			},
		};

		const props = {
			'box-shadow-blur-general': 50,
			'box-shadow-blur-general-hover': 50,
			'box-shadow-blur-unit-general': 'px',
			'box-shadow-blur-unit-general-hover': 'px',
			'box-shadow-horizontal-general': 0,
			'box-shadow-horizontal-general-hover': 0,
			'box-shadow-horizontal-unit-general': 'px',
			'box-shadow-horizontal-unit-general-hover': 'px',
			'box-shadow-inset-general': false,
			'box-shadow-inset-general-hover': false,
			'box-shadow-palette-color-general': 8,
			'box-shadow-palette-color-general-hover': 6,
			'box-shadow-palette-opacity-general': 0.23,
			'box-shadow-palette-opacity-general-hover': 0.23,
			'box-shadow-palette-status-general': true,
			'box-shadow-palette-status-general-hover': true,
			'box-shadow-spread-general': 0,
			'box-shadow-spread-general-hover': 0,
			'box-shadow-spread-unit-general': 'px',
			'box-shadow-spread-unit-general-hover': 'px',
			'box-shadow-status-hover': true,
			'box-shadow-vertical-general': 30,
			'box-shadow-vertical-general-hover': 30,
			'box-shadow-vertical-unit-general': 'px',
			'box-shadow-vertical-unit-general-hover': 'px',
			transition: {
				block: {},
				canvas: {
					border: {
						'transition-duration-general': 0.3,
						'transition-delay-general': 0,
						'easing-general': 'ease',
						'transition-status-general': true,
						hoverProp: 'border-status-hover',
					},
					'box shadow': {
						'transition-duration-general': 0.3,
						'transition-delay-general': 0,
						'easing-general': 'ease',
						'transition-status-general': true,
						hoverProp: 'box-shadow-status-hover',
					},
					'background / layer': {
						'transition-duration-general': 0.3,
						'transition-delay-general': 0,
						'easing-general': 'ease',
						'transition-status-general': true,
						hoverProp: 'block-background-status-hover',
					},
				},
			},
		};

		expect(styleProcessor(obj, {}, props)).toMatchSnapshot();
	});
});
