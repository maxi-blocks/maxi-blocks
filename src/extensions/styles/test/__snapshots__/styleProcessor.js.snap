// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styleCleaner Returns cleaned styles obj 1`] = `
{
  "": {
    "border": {
      "general": {
        "border-bottom-width": "2px",
        "border-color": "var(--maxi-light-color-2)",
        "border-left-width": "2px",
        "border-right-width": "2px",
        "border-style": "solid",
        "border-top-width": "2px",
      },
    },
    "boxShadow": {
      "general": {
        "box-shadow": "0px 0px 0px 0px var(--maxi-light-color-1)",
      },
    },
    "transition": {
      "general": {
        "transition": "box-shadow 0.3s 0s ease",
      },
    },
  },
  " .maxi-container-arrow": {
    "shadow": {
      "general": {
        "filter": "drop-shadow(0px 0px 0px var(--maxi-light-color-1))",
      },
    },
  },
  " .maxi-container-arrow .maxi-container-arrow--content:before": {
    "border": {
      "general": {
        "background": "var(--maxi-undefined-color-2)",
        "height": "calc(50% + 4px)",
        "left": "calc(1px)",
        "top": "calc(1px)",
        "width": "calc(50% + 4px)",
      },
    },
  },
  " > .maxi-background-displayer": {
    "border": {
      "general": {
        "border-bottom-width": "2px",
        "border-color": "var(--maxi-light-color-2)",
        "border-left-width": "2px",
        "border-right-width": "2px",
        "border-style": "solid",
        "border-top-width": "2px",
      },
    },
  },
  ":hover": {
    "border": {
      "general": {
        "border-color": "var(--maxi-light-color-6)",
      },
    },
    "boxShadow": {
      "general": {
        "box-shadow": "5px 5px 0px 0px var(--maxi-light-color-6)",
      },
    },
  },
  ":hover > .maxi-background-displayer": {
    "border": {
      "general": {
        "border-color": "var(--maxi-light-color-6)",
      },
    },
    "size": {
      "general": {
        "transform": "translate(calc(-50% - 0px), calc(-50% - 0px))",
      },
    },
  },
}
`;
